gettext: make `archive.dir.tar` reproducible

raboof opened the pull request:
Fixes #380856
See also https://reproducible-builds.org/docs/archives/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

raboof committed:
gettext: make `archive.dir.tar` reproducible

Fixes #380856 by keeping the original timestamp from the .tar.gz and
explicitly setting file ownership

tobim commented on pkgs/development/libraries/gettext/default.nix:
```diff
@@ -63,7 +63,7 @@ stdenv.mkDerivation rec {
         cp gettext-tools/gnulib-m4/extern-inline.m4 "$oldFilesDir/$oldFile"
       done
-      tar uf gettext-tools/misc/archive.dir.tar -C "$oldFilesDir" "''${oldFiles[@]}"
+      tar uf gettext-tools/misc/archive.dir.tar --sort=name --mtime="@$${SOURCE_DATE_EPOCH}" --owner=0 --group=0 --numeric-owner -C "$oldFilesDir" "''${oldFiles[@]}"
```
This comes out as `-mtime=@1{SOURCE_DATE_EPOCH}'` and tar complains about it:
```
tar: Substituting -9223372036854775807 for unknown date format '@1{SOURCE_DATE_EPOCH}'
```
```suggestion
      tar uf gettext-tools/misc/archive.dir.tar --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0 --numeric-owner -C "$oldFilesDir" "''${oldFiles[@]}"
```

raboof replied:
Oops, botched the escaping there indeed, thanks!

tobim reviewed (changes requested):
This PR causes a lot of rebuilds, so it should target the staging branch instead.
Please follow the instructions in the contributing guide to rebase: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging.

raboof commented:
> This PR causes a lot of rebuilds, so it should target the staging branch instead. Please follow the instructions in the contributing guide to rebase: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging.

Yup, I'm [well-aware](github.com/nixos/nixpkgs/commit/1c73a0773caa843fe16ff6e31eb2fcf7c7362695) of staging, but the job that adds the rebuild stats is still sometimes so slow that by the time they're added I've moved on to other things ;) - thanks for pointing it out, updated the target branch.

raboof commented:
(slightly updated to remove options that don't make sense for `tar uf` and keeping the timestamp from the .tar.gz instead of SOURCE_DATE_EPOCH)
